Fix multiplication of 0-D sparse tensors

pearu opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/65396 and a clang-tidy error.


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #70749



cc @nikitaved @pearu @cpuhrsch

Differential Revision: [D33439136](https://our.internmc.facebook.com/intern/diff/D33439136)

pearu committed:
Fix multiplication of 0-D sparse tensors

[ghstack-poisoned]

pearu commented:
@cpuhrsch , FYI, this is a prerequisite for supporting masked reductions on sparse COO inputs. Details are [here](https://github.com/pytorch/pytorch/pull/64809/files#diff-6d0bd55f94dcbd4ccf83f4006976fa4cb87eb06b72dc16238a21980226be6315R349)

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33439136).

cpuhrsch commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -55,15 +55,21 @@ SparseTensor& mul_out_sparse_zerodim(SparseTensor& r, const SparseTensor& t, con
   AT_ASSERT(value.dim() == 0);
 
+  // Resolve a possibly sparse COO value to a strided tensor.
+  auto& value_ = (value.is_sparse() ? value._values() : value);
```
This raises error

```
error C2440: 'initializing': cannot convert from 'at::Tensor' to 'at::Tensor &'
```

_values() doesn't return a reference. 

As a side note, I personally avoid auto if the resulting type isn't super long otherwise, simply because I can avoid unnecessary type conversions etc., which can be costly. But I also understand if that's not practical or if I'm being overly cautious.

pearu replied:
Fixed. No problem here to avoid auto.

Also enabled a test of multiplying two 0D sparse tensors that hinted that we should also deal with empty sparse tensors here.

pearu committed:
Update on "Fix multiplication of 0-D sparse tensors"

Fixes https://github.com/pytorch/pytorch/issues/65396 and a clang-tidy error.





cc nikitaved pearu cpuhrsch

Differential Revision: [D33439136](https://our.internmc.facebook.com/intern/diff/D33439136)

[ghstack-poisoned]

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33439136).
